File: 44-wildcard-matching/44-wildcard-matching.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        n1 = len(s)
        n2 = len(p)
        dp = [False] * (n2 + 1)
        dp[0] = True
        for idx2 in range (1, n2 + 1):
            allAsterisks = True
            for i in range (1, idx2 + 1):
                if p[i - 1] != '*':
                    allAsterisks = False
                    break
            dp[idx2] = allAsterisks
        for idx1 in range (1, n1 + 1):
            current = [False] * (n2 + 1)
            for idx2 in range (1, n2 + 1):
                if s[idx1 - 1] == p[idx2 - 1] or p[idx2 - 1] == '?':
                    current[idx2] = dp[idx2 - 1]
                elif p[idx2 - 1] == '*':
                    consider = dp[idx2]
                    notConsider = current[idx2 - 1]
                    current[idx2] = consider or notConsider
                else:
                    current[idx2] = False
            dp = current
        return dp[n2]
                    

File: 44-wildcard-matching/44-wildcard-matching.java
class Solution {
    public boolean isMatch(String s, String p) {
        int n1 = s.length(), n2 = p.length();
        boolean[] dp = new boolean[n2 + 1];
        dp[0] = true;
        for(int idx2 = 1; idx2 <= n2; ++idx2) {
            boolean allAsterisks = true;
            for(int i = 1; i <= idx2; ++i) {
                if(p.charAt(i - 1) != '*') {
                    allAsterisks = false;
                    break;
                }
            }
            dp[idx2] = allAsterisks;
        }
        for(int idx1 = 1; idx1 <= n1; ++idx1) {
            boolean[] current = new boolean[n2 + 1];
            for(int idx2 = 1; idx2 <= n2; ++idx2) {
                if(s.charAt(idx1 - 1) == p.charAt(idx2 - 1) | p.charAt(idx2 - 1) == '?')
                    current[idx2] = dp[idx2 - 1];
                else if(p.charAt(idx2 - 1) == '*') {
                    boolean consider = dp[idx2];
                    boolean notConsider = current[idx2 - 1];
                    current[idx2] = consider | notConsider;
                } else current[idx2] = false;
            }
            dp = current;
        }
        return dp[n2];
    }
}